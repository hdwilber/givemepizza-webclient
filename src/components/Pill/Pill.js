import React from 'react'
import PropTypes from 'prop-types'
import _noop from 'lodash/noop'
import styles from './styles.module.css'

class Pill extends React.PureComponent {
  render() {
    const { text, onRemove } = this.props
    return (
      <div className={styles.pill}>
        <span>{ text }</span>
        <span 
          title="remove"
          className={styles.remove}
          onClick={e => {
            e.preventDefault()
            e.stopPropagation()
            onRemove()
          }}
        >
          x
        </span>
      </div>
    )
  }
}
Pill.propTypes = { 
  text: PropTypes.string,
  onRemove: PropTypes.func,
}

Pill.defaultProps = {
  onRemove: _noop,
}

export default Pill
