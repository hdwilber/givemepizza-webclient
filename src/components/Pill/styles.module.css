.pill {
	background-color: gray;
	color: lightgray;
	border-radius: 5px;
}

.remove {
	font-size: 0.5rem;
	display: inline-block;
}
.remove:hover {
	background-color: red;
}
