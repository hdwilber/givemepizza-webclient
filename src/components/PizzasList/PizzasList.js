import React from 'react'
import PropTypes from 'prop-types'
import Pizza from '../Pizza'
import styles from './styles.module.css'

class PizzasList extends React.PureComponent {
  render() {
    const { items, onRemoveItem } = this.props
    return (
      <ul className={styles.list}>
        { items.map(item => {
          const { _id, toppings, name } = item
          const toppingsString = toppings.map(top => top.name).join(',')
          return (
            <li key={_id}>
              {name}
              - Toppings:
              { toppingsString }

              <button onClick={e => { onRemoveItem(item) }}>delete</button>
            </li>
          )
        })
        }
      </ul>
    )
  }
}

PizzasList.propTypes = {
  items: PropTypes.array,
}

PizzasList.defaultProps = {
  items: [],
}

export default PizzasList
