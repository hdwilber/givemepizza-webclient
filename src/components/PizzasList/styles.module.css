.list  {
  list-style: none;
  padding: 0;
  margin: 0;
  display: flex;
  flex-grow: 1;
}

.list li {
  padding: 1rem;
  border: solid 1px;
  display: inline-block;
  margin-right: 1rem;
  position: relative;
}

.list li h1 {
  margin-top: 0;
}
.remove {
  position: absolute;
  top: 0;
  right: 0;

}
