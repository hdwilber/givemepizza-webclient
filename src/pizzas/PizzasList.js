import React from 'react'

class PizzasList extends React.PureComponent {
  render() {
    return (
      <ul>

      </ul>
    )
  }
}
