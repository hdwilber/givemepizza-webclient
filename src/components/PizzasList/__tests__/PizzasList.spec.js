import React from 'react'
import { mount } from 'enzyme'
import PizzasList from '../PizzasList'
import Pill from '../../Pill'

function setupProps() {
  return {
    items: [
      {
        _id: '1',
        name: 'Pizza 1',
        toppings: [{
          _id: 'pill-1',
          name: 'Simple Topping',
        }],
      },
      {
        _id: '2',
        name: 'Pizza 2',
        toppings: [],
      },
      {
        _id: '3',
        name: 'Pizza 3',
        toppings: [],
      },
    ],
    onRemoveItem: jest.fn(),
    onRemoveTopping: jest.fn(),
    onClickItem: jest.fn(),
  }
}

function setup(props) {
  return mount(<PizzasList {...props} />)
}

describe('<PizzasList />', () => {
  let props
  let wrapper

  beforeEach(() => {
    props = setupProps()
    wrapper = setup(props)
  })
  it('should match snapshot', () => {
    expect(<PizzasList {...props} />).toMatchSnapshot()
  })

  it('should render correctly all pizzas', () => {
    const listWrapper = wrapper.find('.list')
    expect(listWrapper.children().length).toEqual(props.items.length)
  })

  it('should call onClickItem when user clicks in the item', () => {
    const index = 0
    const itemWrapper = wrapper.find('li').at(index)
    itemWrapper.simulate('click')
    expect(props.onClickItem).toHaveBeenCalledTimes(1)
    expect(props.onClickItem).toHaveBeenCalledWith(props.items[index])
  })

  it('should call onRemoveItem when user click the remove button', () => {
    const index = 0
    const itemWrapper = wrapper.find('li').at(index)
    const removeButtonWrapper = itemWrapper.find('button')
    removeButtonWrapper.simulate('click', { 
      preventDefault: jest.fn(),
      stopPropagation: jest.fn(),
    })
    expect(props.onRemoveItem).toHaveBeenCalledTimes(1)
  })

  it('should call onRemoveTopping when a topping exists and user clicks in the remove icon of the pill', () => {
    const pillWrapper = wrapper.find(Pill).at(0)
    const removeWrapper = pillWrapper.find('.remove')
    removeWrapper.simulate('click')
    expect(props.onRemoveTopping).toHaveBeenCalledTimes(1)
    expect(props.onRemoveTopping).toHaveBeenCalledWith(props.items[0], props.items[0].toppings[0])
  })
})

