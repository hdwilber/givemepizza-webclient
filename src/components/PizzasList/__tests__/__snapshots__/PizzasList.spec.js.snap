// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PizzasList /> should match snapshot 1`] = `
<PizzasList
  items={
    Array [
      Object {
        "_id": "1",
        "name": "Pizza 1",
        "toppings": Array [
          Object {
            "_id": "pill-1",
            "name": "Simple Topping",
          },
        ],
      },
      Object {
        "_id": "2",
        "name": "Pizza 2",
        "toppings": Array [],
      },
      Object {
        "_id": "3",
        "name": "Pizza 3",
        "toppings": Array [],
      },
    ]
  }
  onClickItem={[MockFunction]}
  onRemoveItem={[MockFunction]}
  onRemoveTopping={[MockFunction]}
/>
`;
